File: topics/Graph/CF580C.py
io = lambda: map(int,input().split())
n,m = io()
p,vis,eg = [0] + list(io()),[0 for _ in range(n+1)],[[] for _ in range(n+1)]
for x,y in [io() for _ in range(n-1)]:
    eg[x].append(y),eg[y].append(x)
cnt = p[1]

def dfs(now,lay):
    if vis[now] :return 
    ok,vis[now] = 0,1
    print("now at {} {}".format(now,lay))
    for i in eg[now]:
        if not vis[i]: 
            dfs(i,0 if (p[i]==0 and lay<=m) else (lay+p[i]))
            ok = 1
    if ok == 0 and lay<=m: 
        cnt+=1
        print("here at {}".format(now))
if __name__ == '__main__':
    dfs(1,p[1])
    print(cnt)

File: topics/Graph/CF_training/CF580C.cpp
/*  -*- coding: utf-8 -*- */
#include <bits/stdc++.h>
#define MEM(a,num) memset(a,num,sizeof(a))
#define rep(i,a,b) for(int i=a;i<=b;i++)
#define fi first
#define pb push_back
#define se second
using namespace std;
typedef long long ll;
typedef pair<int,int> PP;
//const double esp=1e-6
const int maxn  = 5e5;
int vis[maxn],N,M,cnt,p[maxn];
vector<int> v[maxn];
void dfs(int now,int lay)
{
	if(vis[now])return ;
	vis[now] = 1;
	int ok = 0;
	for(int vi:v[now])
		if(!vis[vi])dfs(vi,(lay<=M&&p[vi]==0)?0:lay+p[vi]),ok = 1;
	if(ok==0&&lay<=M)cnt++;
}
int main()
{
	scanf("%d%d",&N,&M);
	int a,b;
	for(int i=1;i<=N;i++)
		scanf("%d",&p[i]);
	for(int i=0;i<N-1;i++){
		scanf("%d%d",&a,&b);
		v[a].pb(b),v[b].pb(a);
	}
	cnt = 0;
	dfs(1,p[1]);
	printf("%d\n",cnt);
}

